test/integ/test_arrow_result.py

```diff
@@ -82,6 +96,87 @@
     "VARCHAR": [None, "ascii_test", b"\xf0\x9f\x98\x80".decode("utf-8")],
 }
 
+# Some datatypes do not match 1:1 with the original data
+PANDAS_REPRS = dict()
+if pandas_available:
+    PANDAS_REPRS = {
+        "TIMESTAMP_LTZ": [
+            pandas.NaT,
+            pandas.Timestamp("2024-01-01 12:00:00+0000", tz="UTC"),
+        ],
+        "TIMESTAMP_NTZ": [pandas.NaT, pandas.Timestamp("2024-01-01 12:00:00")],
+        "TIMESTAMP_TZ": [
+            pandas.NaT,
+            pandas.Timestamp("2024-01-01 12:00:00+0000", tz="UTC"),
+        ],
+        "NUMBER": [numpy.NAN, 1.0, 2.0, 3.0],
+    }
+
+ICEBERG_CONFIG = """
+CATALOG = 'SNOWFLAKE'
+EXTERNAL_VOLUME = 'python_connector_iceberg_exvol'
+BASE_LOCATION = 'python_connector_merge_gate';
+"""
+
+ICEBERG_UNSUPPORTED_TYPES = {
+    "CHAR",
+    "NUMBER",
+    "TIMESTAMP_TZ",
+    "INTEGER",  # Large integer example is incompatible
+}
+
+
+ICEBERG_UNSUPPORTED_NESTED_TYPES = {
+    # SNOW-1320508: Timestamp_LTZ has incorrect scale in iceberg tables
+    "TIMESTAMP_LTZ",
+}
+
+# iceberg testing is only configured in aws at the moment
+ICEBERG_ENVIRONMENTS = {"dev", "aws"}
+ICEBERG_SUPPORTED = os.getenv("cloud_provider", "dev") in ICEBERG_ENVIRONMENTS
+
+PANDAS_UNSUPPORTED_NESTED_TYPES = {
+    # SNOW-1326075: Timestamp types drop information when converted to pandas
+    "TIMESTAMP_LTZ",
+    "TIMESTAMP_NTZ",
+    "TIMESTAMP_TZ",
+    "TIME",
+}
+
+# Generate all valid test cases. By using pytest.param with an id you can
+# run a specific test case easier like so:
+# pytest 'test/integ/test_arrow_result.py::test_dataypes[BINARY-iceberg-pandas]'
+DATATYPE_TEST_CONFIGURATIONS = [
+    pytest.param(
+        datatype,
+        PRIMITIVE_DATATYPE_EXAMPLES[datatype],
+        iceberg,
+        pandas,
+        id=f"{datatype}{'-iceberg' if iceberg else ''}{'-pandas' if pandas else ''}",
+    )
+    for iceberg, pandas, datatype in itertools.product(
+        [True, False],
+        [True, False] if pandas_available else [False],
+        PRIMITIVE_DATATYPE_EXAMPLES,
+    )
+    # Run all tests when not converting to pandas or using iceberg
+    if iceberg is False
+    # Only run iceberg tests on applicable types
+    or (ICEBERG_SUPPORTED and iceberg and datatype not in ICEBERG_UNSUPPORTED_TYPES)
```

Comment:
what's the connector behavior in case of encountering unsupported types?

Reply:
If an unsupported iceberg type is queried then there will be a server-side error from iceberg.

Reply:
gotcha, can we add negative test case for those ICEBERG_UNSUPPORTED_TYPES?